tun/tun_solaris.go

```diff
@@ -0,0 +1,398 @@
+/* SPDX-License-Identifier: MIT
+ *
+ * Copyright (C) 2017-2023 WireGuard LLC. All Rights Reserved.
+ */
+
+package tun
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"sync"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+)
+
+const (
+	ipNode  = "/dev/ip"
+	tunNode = "/dev/tun"
+)
+
+type NativeTun struct {
+	fd          int
+	tunFile     *os.File
+	events      chan Event
+	errors      chan error
+	ipfd        int
+	routeSocket int
+	name        string
+	closeOnce   sync.Once
+}
+
+func (tun *NativeTun) routineRouteListener() {
+	var (
+		statusUp  bool
+		statusMTU int
+	)
+
+	defer close(tun.events)
+
+	iface, err := net.InterfaceByName(tun.name)
+	if err != nil {
+		tun.errors <- err
+		return
+	}
+	tunIfindex := iface.Index
+
+	routeSocket, err := unix.Socket(unix.AF_ROUTE, unix.SOCK_RAW, unix.AF_UNSPEC)
+	if err != nil {
+		tun.errors <- err
+		return
+	}
+	defer unix.Close(routeSocket)
+
+	data := make([]byte, os.Getpagesize())
+	for {
+	retry:
+		n, err := unix.Read(routeSocket, data)
+		if err != nil {
+			if errors.Is(err, unix.EINTR) {
+				goto retry
+			}
+			tun.errors <- err
+			return
+		}
+
+		// You might expect to get a unix.RtMsghdr, but if the Type is RTM_IFINFO then
+		// it's really supposed to be parsed as a IfMsghdr
+		// See https://github.com/illumos/illumos-gate/blob/550b6e4/usr/src/cmd/cmd-inet/usr.sbin/route.c#L2614-L2620
+
+		if uintptr(n) < unsafe.Sizeof(unix.IfMsghdr{}) {
+			continue
+		}
+		msg := (*unix.IfMsghdr)(unsafe.Pointer(&data[0]))
+		if int(msg.Type) != unix.RTM_IFINFO {
+			continue
+		}
+		if int(msg.Index) != tunIfindex {
+			continue
+		}
+
+		iface, err := net.InterfaceByName(tun.name)
+		if err != nil {
+			tun.errors <- err
+			return
+		}
+
+		// Currently the solaris/illumos tun driver has two quirks that should be noted here
+		// 1. At creation time it immediately goes up before this code will have
+		// a chance to notice
+		// 2. It doesn't really respect 'ifconfig tunN down' properly so this code
+		// won't notice, and it will be broken and refuse to come back up
+		// if it is set down.
```

Comment:
Not sure I understand what you mean. It'll get set down but won't notice coming back up? It won't notice going down but it'll come back up just fine?